In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

import seaborn as sns

In [2]:
df = pd.read_csv('train.csv')[
    ['Age','Pclass','SibSp','Parch','Survived']
]
df.head()

Age  Pclass  SibSp  Parch  Survived
0  22.0       3      1      0         0
1  38.0       1      1      0         1
2  26.0       3      0      0         1
3  35.0       1      1      0         1
4  35.0       3      0      0         0

In [3]:
df.isnull().mean()*100

Age         19.86532
Pclass       0.00000
SibSp        0.00000
Parch        0.00000
Survived     0.00000
dtype: float64

In [4]:
# drop the rows which have missing values
df.dropna(inplace=True)

In [5]:
df.head()

Age  Pclass  SibSp  Parch  Survived
0  22.0       3      1      0         0
1  38.0       1      1      0         1
2  26.0       3      0      0         1
3  35.0       1      1      0         1
4  35.0       3      0      0         0

In [6]:
X = df.iloc[:,0:-1]
y = df.iloc[:,-1]

In [7]:
X.head()

Age  Pclass  SibSp  Parch
0  22.0       3      1      0
1  38.0       1      1      0
2  26.0       3      0      0
3  35.0       1      1      0
4  35.0       3      0      0

In [8]:
np.mean(cross_val_score(
    LogisticRegression(),X,y,scoring='accuracy',cv=20
))

np.float64(0.6933333333333332)

**Applying Feature Construction**

In [9]:
X['Family_size'] = X['SibSp'] + X['Parch'] + 1
# the passenger + sibsp + parch
X.head()

Age  Pclass  SibSp  Parch  Family_size
0  22.0       3      1      0            2
1  38.0       1      1      0            2
2  26.0       3      0      0            1
3  35.0       1      1      0            2
4  35.0       3      0      0            1

In [10]:
def myfunc(num):
  if num==1 :
    # alone
    return 0
  elif num > 1 and num <=4 :
    # small family
    return 1
  else:
    # large family
    return 2

In [12]:
myfunc(10)

2

In [13]:
X['Family_type'] = X['Family_size'].apply(myfunc)
X.head()

Age  Pclass  SibSp  Parch  Family_size  Family_type
0  22.0       3      1      0            2            1
1  38.0       1      1      0            2            1
2  26.0       3      0      0            1            0
3  35.0       1      1      0            2            1
4  35.0       3      0      0            1            0

In [14]:
X.drop(columns=[
    'SibSp','Parch','Family_size'
],inplace=True)

In [15]:
X.head()

Age  Pclass  Family_type
0  22.0       3            1
1  38.0       1            1
2  26.0       3            0
3  35.0       1            1
4  35.0       3            0

In [16]:
np.mean(cross_val_score(
    LogisticRegression(),X,y,scoring='accuracy',cv=20
))

np.float64(0.7003174603174602)

**Feature Splitting**

In [17]:
df = pd.read_csv('train.csv')
df.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [18]:
df['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

In [19]:
df['Name'].str.split(',',
                     expand=True)

0                                            1
0       Braund                              Mr. Owen Harris
1      Cumings   Mrs. John Bradley (Florence Briggs Thayer)
2    Heikkinen                                  Miss. Laina
3     Futrelle           Mrs. Jacques Heath (Lily May Peel)
4        Allen                            Mr. William Henry
..         ...                                          ...
886   Montvila                                  Rev. Juozas
887     Graham                         Miss. Margaret Edith
888   Johnston               Miss. Catherine Helen "Carrie"
889       Behr                              Mr. Karl Howell
890     Dooley                                  Mr. Patrick

[891 rows x 2 columns]

In [20]:
df['Name'].str.split(',',
                     expand=True)[1]

0                                  Mr. Owen Harris
1       Mrs. John Bradley (Florence Briggs Thayer)
2                                      Miss. Laina
3               Mrs. Jacques Heath (Lily May Peel)
4                                Mr. William Henry
                          ...                     
886                                    Rev. Juozas
887                           Miss. Margaret Edith
888                 Miss. Catherine Helen "Carrie"
889                                Mr. Karl Howell
890                                    Mr. Patrick
Name: 1, Length: 891, dtype: object

In [21]:
df['Name'].str.split(',',
                     expand=True)[1].str.split(
                         '.',expand=True
                     )

0                                       1     2
0       Mr                             Owen Harris  None
1      Mrs   John Bradley (Florence Briggs Thayer)  None
2     Miss                                   Laina  None
3      Mrs           Jacques Heath (Lily May Peel)  None
4       Mr                           William Henry  None
..     ...                                     ...   ...
886    Rev                                  Juozas  None
887   Miss                          Margaret Edith  None
888   Miss                Catherine Helen "Carrie"  None
889     Mr                             Karl Howell  None
890     Mr                                 Patrick  None

[891 rows x 3 columns]

In [22]:
df['Name'].str.split(',',
                     expand=True)[1].str.split(
                         '.',expand=True
                     )[0]

0         Mr
1        Mrs
2       Miss
3        Mrs
4         Mr
       ...  
886      Rev
887     Miss
888     Miss
889       Mr
890       Mr
Name: 0, Length: 891, dtype: object

In [23]:
df['Title'] = df['Name'].str.split(',',
                     expand=True)[1].str.split(
                         '.',expand=True
                     )[0]
df['Title']

0         Mr
1        Mrs
2       Miss
3        Mrs
4         Mr
       ...  
886      Rev
887     Miss
888     Miss
889       Mr
890       Mr
Name: Title, Length: 891, dtype: object

In [24]:
df[['Title','Name']]

Title                                               Name
0       Mr                            Braund, Mr. Owen Harris
1      Mrs  Cumings, Mrs. John Bradley (Florence Briggs Th...
2     Miss                             Heikkinen, Miss. Laina
3      Mrs       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4       Mr                           Allen, Mr. William Henry
..     ...                                                ...
886    Rev                              Montvila, Rev. Juozas
887   Miss                       Graham, Miss. Margaret Edith
888   Miss           Johnston, Miss. Catherine Helen "Carrie"
889     Mr                              Behr, Mr. Karl Howell
890     Mr                                Dooley, Mr. Patrick

[891 rows x 2 columns]

In [28]:
df.sample(1)

PassengerId  Survived  Pclass  ...        Cabin Embarked  Title
438          439         0       1  ...  C23 C25 C27        S     Mr

[1 rows x 13 columns]

In [33]:
df.groupby('Title')['Survived'].value_counts()

Title         Survived
Capt          0             1
Col           0             1
              1             1
Don           0             1
Dr            0             4
              1             3
Jonkheer      0             1
Lady          1             1
Major         0             1
              1             1
Master        1            23
              0            17
Miss          1           127
              0            55
Mlle          1             2
Mme           1             1
Mr            0           436
              1            81
Mrs           1            99
              0            26
Ms            1             1
Rev           0             6
Sir           1             1
the Countess  1             1
Name: count, dtype: int64

In [41]:
df.shape

(891, 13)

In [44]:
# Mr -> survived
 (81)/(436+81)

0.15667311411992263

In [50]:
df.groupby('Title')['Survived'].count()

Title
Capt              1
Col               2
Don               1
Dr                7
Jonkheer          1
Lady              1
Major             2
Master           40
Miss            182
Mlle              2
Mme               1
Mr              517
Mrs             125
Ms                1
Rev               6
Sir               1
the Countess      1
Name: Survived, dtype: int64

In [51]:
df.groupby('Title')['Survived'].value_counts()/df.groupby('Title')['Survived'].count()

Title         Survived
Capt          0           1.000000
Col           0           0.500000
              1           0.500000
Don           0           1.000000
Dr            0           0.571429
              1           0.428571
Jonkheer      0           1.000000
Lady          1           1.000000
Major         0           0.500000
              1           0.500000
Master        1           0.575000
              0           0.425000
Miss          1           0.697802
              0           0.302198
Mlle          1           1.000000
Mme           1           1.000000
Mr            0           0.843327
              1           0.156673
Mrs           1           0.792000
              0           0.208000
Ms            1           1.000000
Rev           0           1.000000
Sir           1           1.000000
the Countess  1           1.000000
dtype: float64

In [62]:
# proportion of 1's
df.groupby('Title')['Survived'].mean().sort_values(ascending=False)

Title
Lady            1.000000
Ms              1.000000
Sir             1.000000
Mme             1.000000
the Countess    1.000000
Mlle            1.000000
Mrs             0.792000
Miss            0.697802
Master          0.575000
Major           0.500000
Col             0.500000
Dr              0.428571
Mr              0.156673
Capt            0.000000
Jonkheer        0.000000
Don             0.000000
Rev             0.000000
Name: Survived, dtype: float64

people with these saultations had a higher rate of survival :

Mrs	0.792000

Miss	0.697802

Master	0.575000

In [63]:
df.sample(1)

PassengerId  Survived  Pclass  ... Cabin Embarked  Title
888          889         0       3  ...   NaN        S   Miss

[1 rows x 13 columns]

In [64]:
df['Is_Married'] = 0
df.sample(1)

PassengerId  Survived  Pclass  ... Embarked    Title  Is_Married
787          788         0       3  ...        Q   Master           0

[1 rows x 14 columns]

In [74]:
def myfunc(salutation):
  if salutation.strip() == 'Mrs':
    return 1
  else :
    return 0

df['Is_Married'] = df['Title'].apply(myfunc)
df

PassengerId  Survived  Pclass  ... Embarked  Title  Is_Married
0              1         0       3  ...        S     Mr           0
1              2         1       1  ...        C    Mrs           1
2              3         1       3  ...        S   Miss           0
3              4         1       1  ...        S    Mrs           1
4              5         0       3  ...        S     Mr           0
..           ...       ...     ...  ...      ...    ...         ...
886          887         0       2  ...        S    Rev           0
887          888         1       1  ...        S   Miss           0
888          889         0       3  ...        S   Miss           0
889          890         1       1  ...        C     Mr           0
890          891         0       3  ...        Q     Mr           0

[891 rows x 14 columns]

In [70]:
df.head(5)

PassengerId  Survived  Pclass  ... Embarked  Title  Is_Married
0            1         0       3  ...        S     Mr           0
1            2         1       1  ...        C    Mrs           0
2            3         1       3  ...        S   Miss           0
3            4         1       1  ...        S    Mrs           0
4            5         0       3  ...        S     Mr           0

[5 rows x 14 columns]

In [75]:
df['Is_Married']

0      0
1      1
2      0
3      1
4      0
      ..
886    0
887    0
888    0
889    0
890    0
Name: Is_Married, Length: 891, dtype: int64